api/resolver.py

```diff
@@ -180,47 +364,34 @@ class Resolver(object):
 
     Does not tolerate ambiguity at any level of the path except the final node.
     """
+    def __init__(self, id_only=False):
+        self.id_only = id_only 
 
-    @staticmethod
-    def resolve(path):
-
+    def resolve(self, path):
         if not isinstance(path, list):
-            raise Exception("Path must be an array of strings")
-
-        node, resolved, last = Resolver._resolve(path, RootNode)
-        children = node.get_children(last)
+            raise InputValidationException("Path must be an array of strings")
 
-        return {
-            'path': resolved,
-            'children': children
-        }
-
-    @staticmethod
-    def _resolve(path, node, parents=None):
-
-        if parents is None:
-            parents = []
+        path = deque(path)
+        node = None
+        next_node = RootNode()
 
-        last = None
-        if len(parents) > 0:
-            last = parents[len(parents) - 1]
+        resolved_path = []
+        resolved_children = []
 
-        if len(path) == 0:
-            return node, parents, last
+        # Walk down the tree, building path until we get to a leaf node
+        # Keeping in mind that path may be empty
+        while next_node:
+            node = next_node
+            next_node = node.next(path, resolved_path, self.id_only)
 
-        current = path[0]
-        current_id = False
+        # If we haven't consumed path, then we didn't find what we were looking for
+        if len(path) > 0:
+            raise APINotFoundException('Could not resolve node for: ' + '/'.join(path))
 
-        # Check for <id:xyz> syntax
-        if current.startswith('<id:') and current.endswith('>'):
-            current = current[4:len(current)-1]
-            current_id = True
-            print current
+        if hasattr(node, 'get_children'):
+            resolved_children = node.get_children(resolved_path)
 
-        children = node.get_children(last)
-        selected, next_ = node.filter(children, current, current_id)
-
-        path = path[1:]
-        parents.append(selected)
-
-        return Resolver._resolve(path, next_, parents)
+        return {
+            'path': resolved_path,
+            'children': resolved_children
+        }
```

Comment:
Overall, this file roughly doubled in length. It definitely gained some features and documentation, but I think lost a lot of readability along the way. Is there anything we can do to improve that?

Reply:
I think your review had a lot of good suggestions to help with that, but I'll keep an eye out for other improvements that can be made.